### ISSUE TITLE
Protocol for devices to report their configuration / describe are stale

### ISSUE THREAD

**Issue by tacaswell** [bluesky/bluesky#1489]
The root cause of #1487 was that we were caching an invalid result of `obj.describe()` too long.  That patch leverages an existing way to change the configuration to also clear the descriptor cache, but this is less than ideal.

## Expected Behavior

The code in the system that is best position to know when we need to get a new descriptor / configuration is the `ophyd.Device` instances (at the "logic" level using Tom Cobb's wording) because they own the semantics of the device.  We need some protocol for the 

## Current Behavior

Right now plans are responsible for remembering to tell the RE that it needs to clear its caches.  

## Possible Solution

One solution is to push the caching out of the RE and into the devices.  If the caches are in the `Device`, the the RE can call the methods with abandon and minimal performance penalty.  However this would require updating _every_ object to handle its own caching (or pay a big performance cost) and would make it obligatory that every signal involved in configuration or describe be monitored (as there may be no way to force actual call to the control system!).

Another solution would be to take a page from the distributed key-value systems and have an optional `obj.describe_generation : int` (and similar for configuration etc) attribute that anything that wants to cache the results can also cache (with some tricky timing of grab the generation and _then_ grab the values).  On a return call, it can decide if its cache is still good by comparing the generation number (if equal use cache, if greater, re-call).  This puts a bit more complexity in the RE, but lets the reactive callbacks be simpler (they only have to increment the generation, not actually re-compute everything).  This would also open up the possibility of being more aggressive about checking e.g. on every read, that we need to re-generate.  This also has the advantage that not objects will have to change as we can do `generation = getattr(obj, 'describe_generation', 0)` and everything will "just work" which will greatly smooth the adoption path.

It may also be the case that we want to do _both_.  For example, in the case of something like `EpicsMotor` the configuration and result of `obj.describe` changes ~never for many real-world cases, but in the case of many short scans we do 100s of un-needed network calls per run!  To be fair, some of this is mitigated by changing components to trust monitoring.

An idea I rejected is to use a bool as a latch for stale/not stale as the staleness is by the reader (this is related to the issues we have with `collect_asset_docs`) so by using an integer each bit of code that wants to do caching can manage its own latch.

## Context

Right now plans are responsible for remembering to tell the RE that it needs to clear its caches.  This is about the worst of all worlds as the plan is the part of the system that should know the least about when they need to be cleared and it is clearing caches that the plan can only implicitly affect (as the RE decides when to call `describe` and `read_configuration`).

This also leaves us open to issues where the configuration is changed under us (such as by CSS or caput) and we would have no way of recording that.

I could also see this issue being against ophyd, but given that the protocol is between the two and the heavy-lifting of the caching is currently in Bluesky, opened it here.

**Comment by ksunden** [comment]
Such a thing would be useful for this idea of a "hardware proxy server" I'm currently mulling, especially as it extends to pub/sub (I'm leaning towards doing direct RPC first, as that is simpler to integrate into some clients, in particular CLI to simply query a device's current state and exit, but Pub/sub may not be far behind)

**Comment by thomascobb** [comment]
Personally I would prefer to keep the interface as is, but push the caching down to the device level.

The implementation I'm thinking about at the moment for `ophyd.v2` is to cache configuration PV values at stage, then deliver them whenever `describe()` is called.

**Comment by stan-dot** [comment]
@evalott100 isn't this already device layer caching the standard in ophyd-async?

**Comment by evalott100** [comment]
> isn't this the device layer caching already the standard in ophyd-async?

I don't believe we're caching any event-model documents on the device layer?
